### PR TITLE
Add InteliJ debugger work-around

### PR DESCRIPTION
**Pull request by SylvainJuge**
## What does this PR do?

Fixes https://github.com/elastic/apm-agent-java/issues/1673

Idea debugger agent is attached to the JVM and instruments `ForkJoinTask` by modifying the class structure when the `Debug Async Traces` feature is enabled (see related issue).

While this agent declares itself as being able to re-transform classes, it just skips applying bytecode transformation the class is being re-defined, while this optimization makes sense for most applications, it does not when agents are being attached and debugged.

The observed behavior is consistent with [agent code on github](https://github.com/JetBrains/intellij-community/blob/master/java/debugger/debugger-agent/src/com/intellij/rt/debugger/agent/CaptureAgent.java#L168) and the de-compiled agent version shipped with 2020.3

As a result, it makes our agent not able to get the modified bytecode by InteliJ agent, and we try to re-define it with the original class structure, which can't work.
This instrumentation allows to make the InteliJ agent re-generate it's modified version when re-defining classes, which allows our agent to properly instrument `ForkJoinTask`.

## Checklist

- [x] Test locally (was tested on #1305 PR which triggered the warning on agent startup). 
- ~~Find a good way to ensure this advice is applied early and before `ForkJoinTaskInstrumentation`~~
- [x] File an issue or open a PR in Idea repository with a proposed change, maybe adding an extra parameter to avoid this optimization

### CODE REVIEW

**Comment by felixbarny** [comment]
Have you considered creating a PR for the IntelliJ agent that removes `&& classBeingRedefined == null` from the if statement? It seems that would fix the issue. The comment says they don't support retransform but what they actually do is to disallow other agents to retransform, too.

**Comment by SylvainJuge** [comment]
Yes, that's exactly my plan here, (and it's the last item in the checklist).

File: apm-agent-plugins/apm-java-concurrent-plugin/src/main/java/co/elastic/apm/agent/concurrent/BaseInstrumentation.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package co.elastic.apm.agent.concurrent;
+
+import co.elastic.apm.agent.bci.TracerAwareInstrumentation;
+
+public abstract class BaseInstrumentation extends TracerAwareInstrumentation {
+
+    static {
+        if (Boolean.parseBoolean(System.getProperty("intellij.debug.agent"))) {
```
Comment (felixbarny) [inline apm-agent-plugins/apm-java-concurrent-plugin/src/main/java/co/elastic/apm/agent/concurrent/BaseInstrumentation.java:8]:
```suggestion
        if (Boolean.getBoolean("intellij.debug.agent")) {
```

**Review by felixbarny — APPROVED** [review]

**Review by eyalkoren — APPROVED** [review]
Awesome solution for warning only when this instrumentation is applied!

**Comment by SylvainJuge** [comment]
Opened PR on Idea debugger: https://github.com/JetBrains/intellij-community/pull/1528 for a proper fix. In the meantime we have this work-around.

**Pull request merged**